plugins/auth-backend/src/providers/gitlab/provider.ts

```diff
@@ -47,29 +57,58 @@ type PrivateInfo = {
 
 export type GitlabAuthProviderOptions = OAuthProviderOptions & {
   baseUrl: string;
+  signInResolver?: SignInResolver<OAuthResult>;
+  authHandler?: AuthHandler<OAuthResult>;
+  tokenIssuer: TokenIssuer;
+  catalogIdentityClient: CatalogIdentityClient;
+  logger: Logger;
 };
 
-function transformProfile(fullProfile: FullProfile) {
-  const profile = makeProfileInfo({
-    ...fullProfile,
-    photos: [
-      ...(fullProfile.photos ?? []),
-      ...(fullProfile.avatarUrl ? [{ value: fullProfile.avatarUrl }] : []),
-    ],
-  });
+export const extractGitLabUserId = (profile: ProfileInfo): string => {
+  return profile.username || (profile.email?.split('@')[0] as string);
+};
+
+export const gitlabDefaultSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { profile } = info;
 
-  let id = fullProfile.id;
-  if (profile.email) {
-    id = profile.email.split('@')[0];
+  if (!profile.username && !profile.email) {
+    throw new Error('Profile contained no username or email');
   }
 
-  return { id, profile };
-}
+  const id = extractGitLabUserId(profile);
+
+  const token = await ctx.tokenIssuer.issueToken({
+    claims: { sub: id, ent: [`user:default/${id}`] },
+  });
+
+  return { id, token };
+};
+
+export const gitlabDefaultAuthHandler: AuthHandler<OAuthResult> = async ({
+  fullProfile,
+  params,
+}) => ({
+  profile: makeProfileInfo(fullProfile, params.id_token),
+});
 
 export class GitlabAuthProvider implements OAuthHandlers {
   private readonly _strategy: GitlabStrategy;
+  private readonly signInResolver?: SignInResolver<OAuthResult>;
+  private readonly authHandler: AuthHandler<OAuthResult>;
+  private readonly tokenIssuer: TokenIssuer;
+  private readonly catalogIdentityClient: CatalogIdentityClient;
+  private readonly logger: Logger;
 
   constructor(options: GitlabAuthProviderOptions) {
+    this.catalogIdentityClient = options.catalogIdentityClient;
+    this.logger = options.logger;
+    this.tokenIssuer = options.tokenIssuer;
+    this.authHandler = options.authHandler || gitlabDefaultAuthHandler;
+    this.signInResolver = this.createSignInResolverFn(options);
```

Comment:
Let's handle this the same way as it's done [in the other providers](https://github.com/chicoribas/backstage/blob/91a0095f1020db60477fc0495baa98acf10cc74c/plugins/auth-backend/src/providers/microsoft/provider.ts#L297) just to keep things consistent